linkerd/proxy/transport/src/tls/accept.rs

```diff
@@ -32,272 +28,157 @@ pub fn empty_config() -> Arc<Config> {
+    type Io = BoxedIo;
+    type Error = io::Error;
 
-        match &self.tls {
-            // Tls is disabled. Return a new plaintext connection.
+    async fn detect(&self, addrs: Addrs, mut tcp: TcpStream) -> io::Result<(Meta, BoxedIo)> {
+        let local_id = match self.local_identity.as_ref() {
+            Conditional::Some(local_id) => local_id,
             Conditional::None(reason) => {
-                debug!(%reason, "skipping TLS");
                 let meta = Meta {
+                    peer_identity: Conditional::None(reason),
                     addrs,
-                    peer_identity: Conditional::None(*reason),
                 };
-                let conn = (meta, BoxedIo::new(socket));
-                AcceptFuture::accept(self.accept.accept(conn))
+                return Ok((meta, BoxedIo::new(tcp)));
             }
+        };
+
+        let port = addrs.target_addr().port();
+        if self.skip_ports.contains(&port) {
+            debug!(%port, "Skipping TLS detection on port");
+            let meta = Meta {
+                peer_identity: Conditional::None(ReasonForNoPeerName::PortSkipped),
+                addrs,
+            };
+            return Ok((meta, BoxedIo::new(tcp)));
+        }
 
-            // Tls is enabled. Try to accept a Tls handshake.
-            Conditional::Some(tls) => {
-                if self.skip_ports.contains(&target_addr.port()) {
-                    debug!("skipping protocol detection");
-                    let meta = Meta {
-                        peer_identity: Conditional::None(
-                            super::ReasonForNoPeerName::NotHttp.into(),
-                        ),
-                        addrs,
-                    };
-                    let conn = (meta, BoxedIo::new(socket));
-                    AcceptFuture::accept(self.accept.accept(conn))
-                } else {
-                    debug!("attempting TLS handshake");
-                    let meta = AcceptMeta {
-                        accept: self.accept.clone(),
-                        addrs,
-                    };
-                    AcceptFuture::try_tls(TryTls {
-                        meta,
-                        socket,
-                        peek_buf: BytesMut::with_capacity(Self::PEEK_CAPACITY),
-                        config: tls.tls_server_config(),
-                        server_name: tls.tls_server_name(),
-                    })
-                }
+        let no_tls_meta = move |addrs: Addrs| Meta {
+            peer_identity: Conditional::None(ReasonForNoPeerName::NoTlsFromRemote),
+            addrs,
+        };
+
+        // First, try to use MSG_PEEK to read the SNI from the TLS ClientHello.
+        // Because peeked data does not need to be retained, we use a static
+        // buffer to prevent needless heap allocation.
+        //
+        // Anecdotally, the ClientHello sent by Linkerd proxies is <300B. So a
+        // ~500B byte buffer is more than enough.
```

Comment:
It would be interesting to do some actual analysis of this in the wild. We could add a debug metric of the number of times we had to spill over into the fallback buffer or something. Definitely not a blocker, just idle curiosity.

Reply:
We know what the proxy does for its TLS because we can check it. But I don't think the peek approach is reliable if, for instance, only partial data is sent initially. The larger fallback buffer matches the prior behavior as well as the HTTP behavior. I'm not sure what we'd really do with more data. We know this approach helps in the typical case, but there's nothing we can really do about having to support the fallback; and this is already the existing behavior, which works well.

Reply:
Yeah, I think this approach is right. If there isn't any more tuning we can do with additional data, it's definitely not really worth the extra effort, I was just curious.